homework/H2_Linear_Algebra.jl

```diff
@@ -731,8 +732,123 @@ md"How does a high/low bandwidth affect the prediction? How does regularization
 You can also implement [more kernel functions](https://en.wikipedia.org/wiki/Positive-definite_kernel#Examples_of_p.d._kernels) and pass them to the `kernel_ridge` keyword-argument `kernel` to see how the prediction changes.
 "
 
+# ╔═╡ 71c2b3dd-3fb4-4f8b-9e19-a043e05ff498
+md"### Exercise 4 – Multiple dispatch on diagonal matrix"
+
+# ╔═╡ c986580b-fe8c-4b68-9fa1-230b429e6de8
+md"""
+### Overview diagonal matrix
+A diagonal matrix is one that has non-zero elements only in the diagonal.
+
+An example for a diagonal matrix would be:
+"""
+
+# ╔═╡ 57d23d48-4f7b-434c-95f3-dd6c704dbb0b
+A₁ = [3 0 0 
+	  0 7 0 
+	  0 0 -4]
+
+# ╔═╡ d89bd03c-5dda-4141-af90-1673af2370e3
+md"""
+In this typical matrix implementation, all elements are stored, regardless of whether they are zero or non-zero. This results in more storage and computation.
+
+### `Diagonal` from LinearAlgebra.jl
+For that reason, [LinearAlgebra.jl](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/) implements a type called [Diagonal](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.Diagonal) for diagonal matrices. Instead of storing a dense diagonal matrix with all of its off-diagonal zeros, matrices of type `Diagonal` contain only the non-zero diagonal elements in a vector called `diag`:
+
+```julia
+struct Diagonal{V<:AbstractVector} <: AbstractMatrix
+  diag::V
+end
+```
+
+Our previous matrix `A₁` can now be declared as follows: 
+"""
+
+# ╔═╡ 29a57a0e-7859-416b-a6d5-ab8b4ec9516c
+A₂ = Diagonal([3, 7, -4])
+
+# ╔═╡ 3dbc1078-85b1-4711-98fe-29d48d0b1474
+task(
+	md"""
+In this task, you will define two methods for the `trace` function:
+
+1. Implement a method `trace` for inputs of type `AbstractMatrix`
+2. Using multiple dispatch, implement a method `trace` for matrices of type `Diagonal`.
+
+##### Trace Definition:
+
+The trace of an $n \times n$ square matrix $\mathbf{A}$ is defined as:
+
+$\text{trace}(\mathbf{A}) = \sum_{i=1}^{n} a_{ii} = a_{11} + a_{22} + \dots + a_{nn}$
+
+	""", 2
+)
+
+# ╔═╡ 9f99e017-08f6-43b9-9db8-cec9e310b4b7
+function trace(A::AbstractMatrix)  # Don't change this line
+    # Write your code here
+	
+    return 0
```

Comment:
Usually we return `missing` by default.

Reply:
The reason was that I had trouble with the automatic feedback. It gave an error that you cannot compare a boolean with a number. But this is now fixed and will be in the next commit :)